servlet/src/test/java/io/undertow/servlet/test/streams/AbstractServletInputStreamTestCase.java

```diff
@@ -0,0 +1,331 @@
+            urlcon.setRequestProperty("Connection", "close");
+            urlcon.setRequestMethod("POST");
+            urlcon.setDoInput(true);
+            urlcon.setDoOutput(true);
+            OutputStream os = urlcon.getOutputStream();
+            os.write(message.getBytes());
+            os.close();
+            Assert.assertEquals(StatusCodes.OK, urlcon.getResponseCode());
+            InputStream is = urlcon.getInputStream();
+
+            ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+            byte[] buf = new byte[256];
+            int len;
+            while ((len = is.read(buf)) > 0 ){
+                bytes.write(buf, 0, len);
+            }
+            is.close();
+            final String response = new String(bytes.toByteArray(), 0, bytes.size());
+            if (!message.equals(response)) {
+                System.out.println(String.format("response=%s", Hex.encodeHexString(response.getBytes())));
+            }
+            Assert.assertEquals(message, response);
+        } finally {
+            if (urlcon != null) {
+                urlcon.disconnect();
+            }
+        }
+    }
+
+    protected String getBaseUrl() {
+        return DefaultServer.getDefaultServerURL();
+    }
+
+    @Test
+    public void testAsyncServletInputStream3() {
+        String message = "to_user_id=7999&msg_body=msg3";
+        for (int i = 0; i < 200; ++i) {
+            try {
+                runTestViaJavaImpl(message, ASYNC_SERVLET);
+            } catch (Throwable e) {
+                System.out.println("test failed with i equal to " + i);
+                e.printStackTrace();
+                throw new RuntimeException("test failed with i equal to " + i, e);
+            }
+        }
+    }
+
+
+    public void runTest(final String message, String url, boolean preamble, boolean offIOThread) throws IOException {
+        TestHttpClient client = createClient();
+        try {
+            String uri = getBaseUrl() + "/servletContext/" + url;
+            HttpPost post = new HttpPost(uri);
+            if (preamble && !message.isEmpty()) {
+                post.addHeader("preamble", Integer.toString(message.length() / 2));
+            }
+            if (offIOThread) {
+                post.addHeader("offIoThread", "true");
+            }
+            post.setEntity(new StringEntity(message));
+            HttpResponse result = client.execute(post);
+            Assert.assertEquals(StatusCodes.OK, result.getStatusLine().getStatusCode());
+            final String response = HttpClientUtils.readResponse(result);
+            Assert.assertEquals(message.length(), response.length());
+            Assert.assertEquals(message, response);
+        } finally {
+            client.getConnectionManager().shutdown();
+        }
+    }
+
+    public void runTestParallel(int concurrency, final String message, String url, boolean preamble, boolean offIOThread) throws Exception {
```

Comment:
This method is new, invoked by `testAsyncServletInputStreamInParallelOffIoThread` and `testAsyncServletInputStreamInParallel`